File: cpp/1646.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int getMaximumGenerated(int n) {

    if (n == 0)
      return 0;

    ++n;
    vector<int> vec(n, 0);
    vec[1] = 1;

    int ret = 1;
    for (int i = 2; i < n; i++) {
      if (i % 2)
        vec[i] = vec[i / 2] + vec[i / 2 + 1];
      else
        vec[i] = vec[i / 2];
      if (vec[i] > ret)
        ret = vec[i];
    }

    return ret;
  }
};

File: go/1646.go
package main

func getMaximumGenerated(n int) int {
	if n == 0 {
		return 0
	}

	n += 1
	var ret int = 1
	var mem []int = make([]int, n)
  mem[1] = 1
	for i := 2; i < n; i++ {

		if i%2 == 1 {
			mem[i] = mem[i/2] + mem[i/2+1]
		} else {
			mem[i] = mem[i/2]
		}

		if mem[i] > ret {
			ret = mem[i]
		}
	}

	return ret
}
